### ISSUE TITLE
`LoadRequest` and `GetRequest` do not support object equality

### ISSUE THREAD

**Issue by chrisbanes** [coil-kt/coil#405]
```
val request1 = GetRequest.Builder(...).data("blah").build()
val request2 = GetRequest.Builder(...).data("blah").build()
assertEquals(request1, request2) // this fails
```

Can they just be data classes?

**Comment by chrisbanes** [comment]
For context, my integration in [Accompanist](https://github.com/chrisbanes/accompanist) heavily depends on equality checks to only request once per request. I added a very bad workaround in https://github.com/chrisbanes/accompanist/pull/4 which just keys off `data` instead, but that it is obviously not great for complex requests.

**Comment by colinrtwhite** [comment]
I'd want to avoid using data classes for the requests as they create copy and component functions that make it difficult to modify the classes in the future without breaking binary compatibility, however we could auto generate `equals` and `hashCode` implementations.

I've mostly avoided adding them so far since it means custom `Target`, `SizeResolver`, `Transformation`, `Transition`, and `Listener` implementations will need to implement equals as well.

Does Compose overall depend on concrete `equals/hashCode` implementations? Is it possible to use `memoize` instead?

**Comment by chrisbanes** [comment]
We are memoizing, keyed on the `request` instance (via `stateFor(key)`). This enables automatic dropping of the cached result if the request changes. But that requires the key to implement equals/hashCode correctly.

**Comment by colinrtwhite** [comment]
Hmm going to think about this for a bit... We could implement a standard `equals/hashCode` for the request classes - that would at least open the door for users to implement their own `Target`, `SizeResolver`, `Transformation`, `Transition`, and `Listener` `equals/hashCode`. Though, this would also require implementing `equals/hashCode` on `Decoder`s + `Fetcher`s since you can set them on a per-request basis. Internally, we'd basically have to implement `equals/hashCode` for all our components.

**Comment by colinrtwhite** [comment]
This is fixed in `0.12.0` and up as `ImageRequest` (which replaces `LoadRequest` and `GetRequest`) has `equals`/`hashCode`/`toString` implementations.
